### PR TITLE
fix: many namespace lead to provider stuck

### PR DESCRIPTION
**Pull request by shareinto**
### Type of change:

- [x] Bugfix
- [ ] New feature provided
- [ ] Improve performance
- [ ] Backport patches

### What this PR does / why we need it:

Resolve: https://github.com/apache/apisix-ingress-controller/issues/1378
### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [ ] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix-ingress-controller#community) first**

### CODE REVIEW

**Comment by AlinsRan** [comment]
This is not correct, because informer is not started, which will not get any objects.

File: pkg/providers/apisix/provider_init.go [hunk 1]
```diff
@@ -48,23 +48,26 @@ func (p *apisixProvider) Init(ctx context.Context) error {
...
 		log.Debugf("start to watch namespace: %s", key)
 		wg.Add(1)
 		go func(ns string) {
 			defer wg.Done()
 			// ApisixRoute
-			opts := v1.ListOptions{}
 			switch p.common.Config.Kubernetes.APIVersion {
 			case config.ApisixV2beta3:
-				retRoutes, err := p.common.KubeClient.APISIXClient.ApisixV2beta3().ApisixRoutes(ns).List(ctx, opts)
+				retRoutes, err := p.apisixSharedInformerFactory.Apisix().V2beta3().ApisixRoutes().Lister().ApisixRoutes(ns).List(labels.Nothing())
```
Comment (AlinsRan) [inline pkg/providers/apisix/provider_init.go:64]:
```suggestion
				retRoutes, err := p.apisixSharedInformerFactory.Apisix().V2beta3().ApisixRoutes().Lister().ApisixRoutes(ns).List(labels.Everything())
```

File: pkg/providers/apisix/provider_init.go [hunk 2]
```diff
@@ -93,13 +96,13 @@ func (p *apisixProvider) Init(ctx context.Context) error {
 					}
 				}
 			case config.ApisixV2:
-				retRoutes, err := p.common.KubeClient.APISIXClient.ApisixV2().ApisixRoutes(ns).List(ctx, opts)
+				retRoutes, err := p.apisixSharedInformerFactory.Apisix().V2().ApisixRoutes().Lister().ApisixRoutes(ns).List(labels.Nothing())
```
Comment (AlinsRan) [inline pkg/providers/apisix/provider_init.go:99]: Ditto.

File: pkg/providers/apisix/provider_init.go [hunk 3]
```diff
@@ -138,13 +141,13 @@ func (p *apisixProvider) Init(ctx context.Context) error {
 			switch p.common.Config.Kubernetes.APIVersion {
 			case config.ApisixV2beta3:
 				// ApisixConsumer
-				retConsumer, err := p.common.KubeClient.APISIXClient.ApisixV2beta3().ApisixConsumers(ns).List(ctx, opts)
+				retConsumer, err := p.apisixSharedInformerFactory.Apisix().V2beta3().ApisixConsumers().Lister().ApisixConsumers(ns).List(labels.Nothing())
```
Comment (AlinsRan) [inline pkg/providers/apisix/provider_init.go:144]: Ditto.

File: pkg/providers/apisix/provider_init.go [hunk 4]
```diff
@@ -154,13 +157,13 @@ func (p *apisixProvider) Init(ctx context.Context) error {
 					}
 				}
 				// ApisixTls
-				retSSL, err := p.common.KubeClient.APISIXClient.ApisixV2beta3().ApisixTlses(ns).List(ctx, opts)
+				retSSL, err := p.apisixSharedInformerFactory.Apisix().V2beta3().ApisixTlses().Lister().ApisixTlses(ns).List(labels.Nothing())
```
Comment (AlinsRan) [inline pkg/providers/apisix/provider_init.go:160]: Ditto.

File: pkg/providers/apisix/provider_init.go [hunk 5]
```diff
@@ -171,13 +174,13 @@ func (p *apisixProvider) Init(ctx context.Context) error {
 				}
 			case config.ApisixV2:
 				// ApisixConsumer
-				retConsumer, err := p.common.KubeClient.APISIXClient.ApisixV2().ApisixConsumers(ns).List(ctx, opts)
+				retConsumer, err := p.apisixSharedInformerFactory.Apisix().V2().ApisixConsumers().Lister().ApisixConsumers(ns).List(labels.Nothing())
```
Comment (AlinsRan) [inline pkg/providers/apisix/provider_init.go:177]: Ditto.

File: pkg/providers/apisix/provider_init.go [hunk 6]
```diff
@@ -187,13 +190,13 @@ func (p *apisixProvider) Init(ctx context.Context) error {
 					}
 				}
 				// ApisixTls
-				retSSL, err := p.common.KubeClient.APISIXClient.ApisixV2().ApisixTlses(ns).List(ctx, opts)
+				retSSL, err := p.apisixSharedInformerFactory.Apisix().V2().ApisixTlses().Lister().ApisixTlses(ns).List(labels.Nothing())
```
Comment (AlinsRan) [inline pkg/providers/apisix/provider_init.go:193]: Ditto.

**Review by AlinsRan — COMMENTED** [review]
Sorry for the delay.
I thinking using `Labels.Nothing()` is not correct. This will not get the object.
https://pkg.go.dev/k8s.io/apimachinery/pkg/labels#Nothing

**Comment by AlinsRan** [comment]
We have a lot of `informer.run()`, `cache.wait()` in our code, actually it only needs to call factor.start() and `factor.wait()` to complete.
There is a conflict between them and they can only be started once. I suggest that the codes of `informer.run` and `cache.wait()` should be deleted. 
What do you think?

File: pkg/providers/apisix/provider_init.go (hunk 1 above)
Comment (shareinto) [reply pkg/providers/apisix/provider_init.go]: Thanks for reminding, I'll fix it as soon as possible

File: pkg/providers/apisix/provider_init.go [hunk 7]
```diff
@@ -48,23 +48,26 @@ func (p *apisixProvider) Init(ctx context.Context) error {
 		pluginConfigMapA6 = make(map[string]string)
 	)
 
+	p.apisixSharedInformerFactory.Start(ctx.Done())
+	p.apisixSharedInformerFactory.WaitForCacheSync(ctx.Done())
```
Comment (tao12345666333) [inline pkg/providers/apisix/provider_init.go:52]: I think we should check its return value, and log it.

```go
	synced := p.apisixSharedInformerFactory.WaitForCacheSync(ctx.Done())

	for v, ok := range synced {
		if !ok {
			// Log it and return
		}
	}


```

**Comment by tao12345666333** [comment]
> We have a lot of `informer.run()`, `cache.wait()` in our code, actually it only needs to call factor.start() and `factor.wait()` to complete.
> 
> There is a conflict between them and they can only be started once. I suggest that the codes of `informer.run` and `cache.wait()` should be deleted. 
> 
> 
> 
> What do you think? 
> 
> 
> 
> cc @tao12345666333 

Sorry, I forgot to reply you.  I think so.  Meanwhile, I think it can be done in a follow-up PR

**Review by lingsamuel — APPROVED** [review]

**Review by tao12345666333 — APPROVED** [review]

**Review by AlinsRan — APPROVED** [review]

**Pull request merged**